### DISCUSSION TITLE
out ref parameters

### DISCUSSION THREAD

**Discussion by bill-poole** [dotnet/csharplang#4152] [Language Ideas]
C# currently has the ability for a method to return a "ref" or "ref readonly" value, but no such ability exists with "out" parameters. That is, I'd like the ability to specify a "ref" or "ref readonly" out parameter of a method. I have an example below where I would use this feature if available. I've only included the pertinent parts from the code as I'd write it if the feature were available.

```csharp
public class ApportionedValue
{
  private Portion[] _portions;

  public int UnspecifiedPortionIndex { get; } = -1;

  public bool HasUnspecifiedPortion => UnspecifiedPortionIndex >= 0;

  public bool IsPartlySpecifiedNumeric(out ref readonly Portion specifiedPortion)
  {
    specifiedPortion = ref Portion.Empty;
    if (_portions.Length != 2)
    {
      return false;
    }

    if (!HasUnspecifiedPortion)
    {
      return false;
    }

    ref readonly var portion = ref _portions[1 - UnspecifiedPortionIndex];
    if (portion.Value.IsNumeric)
    {
      specifiedPortion = ref portion;
      return true;
    }

    return false;
  }

  // ...
}

public struct Value
{
  public bool IsNumeric { get; }

  // ...
}

public struct Portion
{
  public static readonly Portion Empty;

  public Value Value { get; }

  // ...
}
```

I'd be keen to get people's thoughts.

**Comment by HaloFour** [comment]
I'm not sure I understand the purpose of this proposal.  Are you trying to achieve a `ref` to a `ref`?

**Reply by bill-poole** [reply]
No, I want the IsPartlySpecifiedNumeric method to return 'true' if the portions in the _portions array field meet the criteria of the ApportionedValue being a "partly-specified numeric" (which is an ApportionedValue with exactly two portions, where one portion is the specified portion and the other portion is the unspecified portion) and have the out parameter be a ref to the specified portion, which is one of the elements in the _portions array.

If I were to do this as a regular out parameter (i.e. "out Portion portion" instead of "out ref readonly Portion portion", then the specified portion element in the _portions array would be copied into the out parameter, rather than just a reference to that portion.

**Reply by HaloFour** [reply]
`out` already means `ref`, just with additional rules that the compiler assign a value.

**Reply by bill-poole** [reply]
But isn't it a ref to a value (i.e. a Portion instance) that has been allocated on the stack prior to the method being invoked? I don't want an empty Portion instance to be allocated on the stack and then copied into by the method from the relevant element in the _portions array. I want the out value to be a 64-bit pointer (on a 64-bit system) to the relevant element in the _portions array.

**Comment by YairHalberstadt** [comment]
Out is always passed by ref.

**Reply by bill-poole** [reply]
Yes that's true. But isn't it implemented by the compiler as a new Portion instance on the stack into which the out value is copied? That is, if I have some method MyMethod(out MyStruct x), then a MyStruct instance will be allocated on the stack, into which the value will then be written.

I'm seeking for a ref pointer to be allocated on the stack, into which a ref to a Portion (or in this example MyStruct) instance would be written.

So in one case, the full MyStruct (or Portion) struct size is allocated on the stack, into which the full instance is copied - as opposed to just a ref (64 bits on a 64-bit system).

**Reply by HaloFour** [reply]
I think that would require a concept of `ref ref`, which seems supported by the runtime but the compiler has no concept of it.

**Reply by bill-poole** [reply]
Well, it doesn't seem too different from ref locals, which we already have. For example, we can do the following:

```csharp
public void Whatever()
{
  var x = 0;
  ref readonly var y = ref x;
}
```

It kind of looks like the variable y is a ref to a ref, but it's actually just a ref to x. I'm thinking the same kind of mechanism might work here.

**Reply by bill-poole** [reply]
I would have thought it would be implementable using the exact same underlying mechanism used to have a "ref return" or "ref readonly return" method - except the return value would be placed in an out parameter. The out parameters and return value are all allocated on the stack. Out parameters are just extra return values under the hood, aren't they?

**Reply by HaloFour** [reply]
`ref` locals and `ref` parameters are references to a memory location containing the value.  As an `out` is already a `ref` an `out ref` would be a reference to a location containing a reference to another location containing the value.  It wouldn't be the same mechanism as it's a different data type entirely, e.g. an `int32&&` vs. an `int32&`.

**Reply by bill-poole** [reply]
Sure, but I'm not proposing that the "out ref" syntax would actually mean a ref to a ref. That is, even though an "out" parameter is already a ref (to a value allocated on the stack prior to the method's invocation), we wouldn't be defining an "out ref" to be a ref to a ref. That is, it would be operate very similarly to the way that with the "ref locals" language feature we have a ref to a ref variable meaning a ref to the actual variable, not a ref to a ref.

Today with we can have a method MyMethod(out MyStruct myStruct), where an instance of MyStruct is allocated on the stack prior to the method being invoked. When the method writes to the myStruct parameter, it is writing to this variable  - which in effect is just like any other variable, but allocated on the stack by the caller and available to the caller after the method returns.

I'm proposing that an "out ref" or "out ref readonly" parameter would in effect act as a "ref local" variable allocated on the stack prior to the method's invocation, which would then be written to by the method just the same as the myStruct variable and just like the ordinary "ref local" mechanism that exists today.

That is, the "out ref"/"out ref readonly" paremeters would be just ref locals like any other ref local - i.e. not a ref to a ref.

**Reply by HaloFour** [reply]
It seems that a `ref ref` isn't technically legal today per the [CLR ECMA spec](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf):

> II.14.4.2
> A managed pointer cannot point to another managed pointer, but it can point to an object reference or a value type.

So such a feature might require runtime work as well, or the spec needs to be updated.

**Reply by bill-poole** [reply]
But I'm not proposing a ref to a ref, which I explained in my previous reply. This proposal doesn't involve nor require a ref to a ref in any way. The "out ref"/"out ref readonly" parameters would themselves be managed pointers to an object reference or value type - stipulated as permitted by the clause you cited from the CLR ECMA spec.

**Reply by HaloFour** [reply]
@billpoole-mi 

I'm not sure I understand your distinction.  An `out` parameter is already a `ref` parameter which behaves as a `ref` local.  When calling a method the `ref` is passed by value and cannot be reassigned.  That's not possible without passing a `ref` to that `ref`.

**Reply by bill-poole** [reply]
I'll explain the distinction by way of an example. Consider the following code.

```csharp
public class Whatever
{
  private readonly MyStruct[] _values = new MyStruct[1];

  public void GetValue(out MyStruct value)
    => value = _values[0];
}
```

In the above code, the MyStruct "value" parameter is passed by reference as you say. But it is a reference to a MyStruct instance allocated on the stack by the caller prior to invoking the GetValue method such that the value is available to the caller after the method completes. The first element in the _values array is _copied_ into the out parameter (in its entirety, as opposed to a pointer to the first element).

Now consider this code.

```csharp
public class Whatever
{
  private readonly MyStruct[] _values = new MyStruct[1];

  public void GetValue(out ref readonly MyStruct value)
    => value = ref _values[0];
}
```

In this code, a _reference_ to the first element of the _values array is copied into the out parameter. i.e. in this case, the "value" out parameter is a _managed pointer_; not a MyStruct instance. Conversely in the first example, the "value" out parameter is a MyStruct instance; not a managed pointer.

**Reply by HaloFour** [reply]
In both cases the parameter is a managed reference.  That's what `out` parameters are, `ref` parameters.

If your desire is to be able to reassign `value` so that the caller would then be able to observe the value at `_values[0]` that would require a `ref` to a `ref`.

**Reply by bill-poole** [reply]
Okay, yes good point. In both cases the "value" out parameter is a managed pointer. But the difference is what the managed pointer points to. In the first example, the managed pointer points to a MyStruct instance allocated on the stack by the caller and setting that parameter value involves _copying_ the first element of the array into that referenced memory location.

In the second example, we'd just be updating the pointer itself. So actually (and thank you for correcting me) the caller does the same thing in both cases - allocates a managed pointer on the stack. But, in the first example, the method implementation writes a copy of the first element of the _values array into the address referenced by that pointer. And in the second example, the method implementation updates the pointer - i.e. it doesn't write anything into the address referenced by that pointer.

**Reply by bill-poole** [reply]
Actually let me clarify a bit more. In the first example, the caller allocates a MyStruct instance on the stack prior to invoking the method and sets the "value" out parameter to be a managed pointer to that MyStruct instance. When the method implementation sets that parameter, it _copies_ the first element of the array into that referenced memory location.

In the second example, the pointer doesn't actually point to anything (i.e. it is uninitialized) when the method is invoked. That is, the caller allocates an uninitialized managed pointer to a MyStruct instance on the stack. The method implementation then updates that pointer to point to the first element of the _values array.

**Reply by HaloFour** [reply]
> In the second example, we'd just be updating the pointer itself.

And that's what requires a `ref` to a `ref`.  You are reassigning the pointer, but that reassignment is lost because the pointer is passed by value.  You can see this behavior with this example:

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQuAzAAQFUDCVA3mVa1QA4ICWAbhPDUIA2Kl0wwA2gF0qfADbo4UKgF4qmOAHdR46UyqEkVfEeoBfANws211pUEjcAFioBZABQI4AMx0wqAB4AlEy2bKwBqlRevvKKUBIADFJWpOFmYWG4wjQurhBi7iHMaeFsYv6RaompZWweMYFBtXWCAJzuwS2sGaRmQA)

**Reply by bill-poole** [reply]
But hang on, today we already have "ref returns" methods. i.e. we can have a method like below.

```csharp
public ref readonly MyStruct GetValue()
  => ref _values[0];
```

In this case, the caller allocates a managed pointer on the stack, which is where the returned value will be written. The method then writes the reference to the first element of the array to this managed pointer.

What we're talking about here is the exact same thing - but syntactically expressed as an out parameter rather than a return value.

If it doesn't require a ref to a ref to write the managed pointer to the return value (on the stack), then why would it require a ref to a ref to write the managed pointer to an out parameter (also on the stack)?

**Reply by YairHalberstadt** [reply]
> In this case, the caller allocates a managed pointer on the stack, which is where the returned value will be written. The method then writes the reference to the first element of the array to this managed pointer.

That's not how ref returns works. They literally just returns the ref directly (a ref is ultimately just an integer)

**Reply by bill-poole** [reply]
Yes, but that ref return value (an integer as you say) is written somewhere so the caller can read it - i.e. it is written to the stack. I'm proposing the exact same thing here to work in the exact same way. "out" parameters are just extra return values, are they not?

That is, the following code is valid today.

```csharp
public ref readonly int GetValue(int[] values)
    => ref _values[0];
```

Under this proposal, one could achieve the exact same outcome, just with a different C# syntax.

```csharp
public void GetValue(int[] values, out ref readonly int value)
    => value = ref _values[0];
```

So the compiler would emit very similar bytecode in both cases. This proposal is just extending the power of ref returns to out parameters. Is there some underlying, fundamental difference between return values and "out" parameters in the mechanics of the runtime I'm missing?

**Reply by HaloFour** [reply]
@billpoole-mi 

Parameters are pushed by the caller, return values are pushed by the callee.  The callee cannot overwrite the parameters pushed by the caller.  You cannot treat a `ref` return as equivalent to a `ref ref` parameter because they are not the same thing to the runtime.

**Reply by bill-poole** [reply]
Ah, bummer. That's what I was missing. Thus the need for a ref to a ref, which you were saying all along. I was blind, but I now see!

Is it a runtime restriction that the callee cannot overwrite the parameters pushed by the caller? Or has there just been no need for it in the past? It doesn't seem like it would be breaking any memory safety rules to write values pushed by the caller that were intended by the caller to be written by the callee.

**Reply by bill-poole** [reply]
And another thought, would it be acceptable (i.e. would it be supported in the runtime) for "out ref" and "out ref readonly" parameters to be pushed by the callee, along with the return value? i.e. would the runtime currently support the callee pushing more than just the return value, such that "out ref" and "out ref readonly" parameters would then be literally just extra return values? Or would that require a runtime change?

**Reply by CyrusNajmabadi** [reply]
There would have to be an enormously compelling use case for this, likely with a very large customer (or customers) asking for it.  Can thse not just be returned from the method instead of trying to do them as out-parameters?

**Reply by YairHalberstadt** [reply]
> Can thse not just be returned from the method instead of trying to do them as out-parameters?

Without ref type parameters you need to return a custom ref struct, instead of a tuple of refs.

**Reply by HaloFour** [reply]
To my understanding there is no IL that would support it.  When the caller pushes the `ref` onto the evaluation stack it pushes a copy.  By the time the method returns the stack has been popped from the evaluation stack, so even if the callee overwrote the arg there's nothing for the caller to read back out.

While the spec does state that managed pointers can't point to other managed pointers it's possible (even likely) that the runtime technically supports it.  The ECMA spec predates the addition of `ref` returns/locals.  But even if they possible the question might be whether adding them is worthwhile given that much of the work around `ref` returns/locals and `ref` reassignment was about the compiler enforcing that these operations can be done safely outside of an `unsafe` context.

**Reply by bill-poole** [reply]
Well, there are a couple of issues working within that constraint. The first is that we're constrained to a single ref return value - so it wouldn't be possible to support the use case of having multiple "out ref"/"out ref readonly" parameters - unless I'm missing something?

But then there's also the issue (arguably much less important) of the syntax ugliness of the following.

```csharp
public ref readonly Portion IsPartlySpecifiedNumeric(out bool isPartlySpecifiedNumeric)
{
  if (_portions.Length != 2)
  {
    isPartlySpecifiedNumeric = false;
    return ref Portion.Empty;
  }

  if (!HasUnspecifiedPortion)
  {
    isPartlySpecifiedNumeric = false;
    return ref Portion.Empty;
  }

  ref readonly var portion = ref _portions[1 - UnspecifiedPortionIndex];
  if (portion.Value.IsNumeric)
  {
    isPartlySpecifiedNumeric = true;
    return ref portion;
  }

  isPartlySpecifiedNumeric = false;
  return ref Portion.Empty;
}
```

Which would be invoked as follows.

```csharp
var specifiedPortion = apportionedValue.IsPartlySpecifiedNumeric(out var isPartlySpecifiedNumeric);
if (isPartlySpecifiedNumeric)
{
  // Do something with specifiedPortion
}
```

i.e. it would be much cleaner syntax to call as follows.

```csharp
if (apportionedValue.IsPartlySpecifiedNumeric(out readonly ref var specifiedPortion)
{
  // Do something with specifiedPortion
}
```

I completely agree that this doesn't meet the bar for a runtime change - if one is required, which it is sounding like it is.

**Reply by bill-poole** [reply]
Okay thanks @HaloFour and @YairHalberstadt. You've both been extremely helpful and supportive discussing this proposal - it's very much appreciated.

So, just to summarise and make sure I've understood correctly - it would be a byte code and/or runtime change to support this proposal. And it certainly doesn't meet the bar for that kind of change.

**Reply by bill-poole** [reply]
> Without ref type parameters you need to return a custom ref struct, instead of a tuple of refs.

Also, you couldn't achieve this by returning a custom ref struct because ref struct can't (yet) contain ref fields; although there is currently a proposal to do so.

**Reply by CyrusNajmabadi** [reply]
> The first is that we're constrained to a single ref return value -

Yup.  but i'm curious how essential it is that this be multiple values, as opposed to multple calls that could each return one value.  In other words, if there is a suitable workaround, it vastly diminishes the need for something so specialized.  Of course, i get why it would still be unpalatable to you, but this would still weigh into the decision making.

**Reply by CyrusNajmabadi** [reply]
> Also, you couldn't achieve this by returning a custom ref struct because ref struct can't (yet) contain ref fields; although there is currently a proposal to do so.

Right.  That would likely be the preferred way for this to go for us.  this use case woudl then just fall out cleanly.

**Reply by bill-poole** [reply]
Well, you could have a method that does something rather expensive and the output of that method is two or more "ref return" outputs. If you did that as two calls, then you'd need to do that rather expensive thing twice.

But let me say, the ugly syntax I mentioned isn't my primary motivator here. Ultimately when I'm designing/writing high performance code and I establish a way that it should _theoretically_ be possible to work, but the C# language syntax or class libraries get in the way, it introduces unnecessary overheads - typically in the form of unnecessary copying of memory.

That as I understand it was the motivation for the wonderful additions to the language and class libraries in recent years - making specific reference to the features added in C# 7, 7.1, 7.2 and 7.3, and also Span<T>, etc. It's been a great journey with a lot of progress. I just hope we one day reach a point with the language and class libraries where we have the expressiveness we need to direct the machine to do what we want/need without constraint (without threatening memory safety that is).

**Reply by bill-poole** [reply]
> Right. That would likely be the preferred way for this to go for us. this use case woudl then just fall out cleanly.

Yes, this would (hopefully) allow the following solution (or something along these lines - apologies, I'm not really fully up to speed on the current state of the proposal for ref fields in ref structs).

```csharp
public ref struct PartlySpecifiedNumeric
{
  public readonly ref Portion SpecifiedPortion;

  public PartlySpecifiedNumeric(ref readonly Portion specifiedPortion)
  {
    SpecifiedPortion = ref specifiedPortion;
  }
}

public bool IsPartlySpecifiedNumeric(out PartlySpecifiedNumeric partlySpecifiedNumeric)
{
  partlySpecifiedNumeric = new PartlySpecifiedNumeric(ref Portion.Empty);
  if (_portions.Length != 2)
  {
    return false;
  }

  if (!HasUnspecifiedPortion)
  {
    return false;
  }

  ref readonly var portion = ref _portions[1 - UnspecifiedPortionIndex];
  if (portion.Value.IsNumeric)
  {
    partlySpecifiedNumeric = new PartlySpecifiedNumeric(ref portion);
    return true;
  }

  return false;
}
```

This would then be invoked as follows:

```csharp
if (apportionedValue.IsPartlySpecifiedNumeric(out var partlySpecifiedNumeric)
{
  // Do something with partlySpecifiedNumeric.SpecifiedPortion
}
```

It's not really clear to me what the ref fields of a ref struct that can contain ref fields would be initialised to for a default instance of the ref struct. Maybe the compiler will prevent that from happening?

**Reply by bill-poole** [reply]
Thanks also to @CyrusNajmabadi for your help with this.

**Reply by bill-poole** [reply]
Would it be possible to use the (currently internal) `ByReference<T>` type to support this proposal? For example, consider the following code.

```csharp
public void MyMethod(out ref MyStruct x)
  => ref _values[0];
```

Could that be implemented using `ByReference<T>` as follows?

```csharp
public void MyMethod(out ByReference<MyStruct> x)
  => new ByReference<MyStruct>(ref _values[0]);
```

**Reply by bill-poole** [reply]
By way of example, the following is valid in C# today.

```csharp
public readonly ref struct ByReference<T>
{
    private static T Null;

    private readonly Span<T> _span;

    public readonly ref T Value
        => ref (_span.IsEmpty ? ref Null : ref _span[0]);

    public ByReference(ref T value)
    {
        _span = MemoryMarshal.CreateSpan(ref value, 1);
    }
}

public readonly struct MyStruct
{
}

public class Whatever
{
    private MyStruct[] _values = new MyStruct[1];

    public void WriteRefToOutParameter(out ByReference<MyStruct> value)
        => value = new ByReference<MyStruct>(ref _values[0]);

    public void UseRefInParameter(in MyStruct value)
    {
        // Do something with value
    }

    public void GetRefOutParameterAndPassAsInParameter()
    {
        ReturnRefInOutParameter(out var valueRef);
        UseRefInParameter(valueRef.Value);
    }
}
```
